File: _stats_build/statistics/wr_bpa.rb
require_relative "abstract/round_metric"

class WrBpa < RoundMetric
  def initialize
    @title = "BPA (Best Possible Average)"
    @title_zh = "BPA（最佳可能平均）"
    @note = "Best Possible Average: average of the best 3 out of the first 4 solves in a round."
    @note_zh = "BPA：一轮中前 4 次成绩取最好的 3 次计算平均。"
    @table_header = { "Result" => :right, "Improvement" => :right, "Days" => :right, "Person" => :left, "Competition" => :left, "Date" => :left, "Details" => :left }
  end

  # NOTE: BPA = 前4次中取最好的3次求均值
  # 如果前4次中有2个及以上 DNF/DNS，则无效
  def compute_metric(values, r)
    first4 = values[0..3]
    valid = first4.select { |v| v > 0 }
    return nil if valid.length < 3
    # 取最好的3个（最小值）
    best3 = valid.sort.first(3)
    best3.sum.to_f / 3
  end
end

File: core/packages/stats-build/src/statistics/wr_bpa.ts
// NOTE: BPA (Best Possible Average)——前 4 次中取最好 3 次均值
import { RoundMetric } from '../core/round_metric.js';
import { EVENTS_WITH_AO5 } from '../core/events.js';

export class WrBpa extends RoundMetric {
  constructor() {
    super();
    this.title = 'BPA';
    this.titleZh = 'BPA';
    this.note = 'Best Possible Average: average of the best 3 out of the first 4 solves in a round.';
    this.noteZh = '最佳可能平均：前 4 次中取最好 3 次的平均。';
    this.tableHeader = {
      'Result': 'right', 'Improvement': 'right', 'Days': 'right',
      'Person': 'left', 'Date': 'left', 'Competition': 'left', 'Details': 'left',
    };
  }

  targetEvents() { return EVENTS_WITH_AO5; }

  // NOTE: BPA = 前 4 次中取最好 3 次均值
  computeMetric(values: number[]): number | null {
    const first4 = values.slice(0, 4);
    const valid = first4.filter(v => v > 0);
    if (valid.length < 3) return null;
    const best3 = valid.sort((a, b) => a - b).slice(0, 3);
    return best3.reduce((s, v) => s + v, 0) / 3;
  }
}
